observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerSubsystemRegistrar.java

```diff
@@ -0,0 +1,189 @@
+        ManagementResourceRegistration registration =
+                parent.registerSubsystemModel(ResourceDefinition.builder(ResourceRegistration.of(PATH), RESOLVER).build());
+        ResourceDescriptor descriptor = ResourceDescriptor.builder(RESOLVER)
+                .withRuntimeHandler(ResourceOperationRuntimeHandler.configureService(this))
+                .addCapability(MICROMETER_COLLECTOR_RUNTIME_CAPABILITY)
+                .addCapability(MICROMETER_REGISTRY_RUNTIME_CAPABILITY)
+                .addAttributes(List.of(ENDPOINT, STEP, EXPOSED_SUBSYSTEMS))
+                .build();
+
+        ManagementResourceRegistrar.of(descriptor).register(registration);
+
+        return registration;
+    }
+
+    @Override
+    public ResourceServiceInstaller configure(OperationContext context, ModelNode model) throws OperationFailedException {
+        List<String> exposedSubsystems = MicrometerSubsystemRegistrar.EXPOSED_SUBSYSTEMS.unwrap(context, model);
+        boolean exposeAnySubsystem = exposedSubsystems.remove("*");
+        String endpoint = MicrometerSubsystemRegistrar.ENDPOINT.resolveModelAttribute(context, model)
+                .asStringOrNull();
+        Long step = MicrometerSubsystemRegistrar.STEP.resolveModelAttribute(context, model)
+                .asLong();
+
+        WildFlyMicrometerConfig config = new WildFlyMicrometerConfig(endpoint, step);
+        wildFlyRegistry = new WildFlyOtlpRegistry(config);
+
+        context.addStep(new AbstractDeploymentChainStep() {
+            @Override
+            public void execute(DeploymentProcessorTarget processorTarget) {
+                processorTarget.addDeploymentProcessor(MicrometerConfigurationConstants.NAME, DEPENDENCIES,
+                        DEPENDENCIES_MICROMETER, new MicrometerDependencyProcessor());
+                processorTarget.addDeploymentProcessor(MicrometerConfigurationConstants.NAME, POST_MODULE,
+                        POST_MODULE_MICROMETER,
+                        new MicrometerDeploymentProcessor(exposeAnySubsystem, exposedSubsystems, wildFlyRegistry));
+            }
+        }, RUNTIME);
+
+
+        ServiceDependency<ModelControllerClientFactory> mccf = ServiceDependency.on(CLIENT_FACTORY_CAPABILITY, ModelControllerClientFactory.class);
+        ServiceDependency<Executor> executor = ServiceDependency.on(MANAGEMENT_EXECUTOR, Executor.class);
+        ServiceDependency<ProcessStateNotifier> processStateNotifier =
+                ServiceDependency.on(PROCESS_STATE_NOTIFIER, ProcessStateNotifier.class);
+
+        Supplier<MicrometerCollector> collectorSupplier = () ->
+                new MicrometerCollector(mccf.get().createClient(executor.get()), processStateNotifier.get(),
+                        wildFlyRegistry);
+
+        context.addStep((operationContext, modelNode) -> {
+            MicrometerCollector micrometerCollector = collectorSupplier.get();
```

Comment:
This doesn't look right.  This will cause a second collector to be created.
Rather, you want to capture the value created by the service.

Reply:
I had
`
MicrometerCollector micrometerCollector = 
    (MicrometerCollector) operationContext.getServiceRegistry(false)
    .getRequiredService(MICROMETER_COLLECTOR_RUNTIME_CAPABILITY
    .getCapabilityServiceName()).getValue();
`
but I got an NPE. Is it ok to cache in the Supplier and do a null-check there, or is there a Better Way?

Reply:
How's that?

Reply:
> MicrometerCollector micrometerCollector =  (MicrometerCollector) operationContext.getServiceRegistry(false) .getRequiredService(MICROMETER_COLLECTOR_RUNTIME_CAPABILITY .getCapabilityServiceName()).getValue();

This is incorrect.  The installed service does not implement org.jboss.msc.Value.getValue() and thus will always return null.

Reply:
> Is it ok to cache in the Supplier and do a null-check there, or is there a Better Way?

```java
AtomicReference<MicrometerCollector> captor = new AtomicReference<>();

context.addStep((operationContext, modelNode) -> {
        MicrometerCollector collector = captor.get();
        // Given that this step runs in the VERIFY stage, and our collector service was started eagerly, the collector reference _should_ be non-null.
        if (collector == null) return;
        // ...
    }, OperationContext.Stage.VERIFY);

return CapabilityServiceInstaller.builder(MICROMETER_COLLECTOR_RUNTIME_CAPABILITY, collectorSupplier)
        .requires(List.of(mccf, executor, processStateNotifier))
        .withCaptor(captor::set) // capture the provided value
        .asActive() // Start actively
        .build();
```